File: intermediate_data_structures/intermediate_ds/graph.py
class Graph:
    def __init__(self, directed=False): self.directed = directed; self._edges = {}
    def add_vertex(self, vertex): self._edges.setdefault(vertex, set())
    def add_edge(self, source, target):
        self.add_vertex(source); self.add_vertex(target); self._edges[source].add(target)
        if not self.directed: self._edges[target].add(source)
    def bfs(self, start):
        if start not in self._edges: raise KeyError(start)
        seen, order, pending = {start}, [], [start]
        while pending:
            vertex = pending.pop(0); order.append(vertex)
            for neighbor in sorted(self._edges[vertex]):
                if neighbor not in seen: seen.add(neighbor); pending.append(neighbor)
        return order

File: go_data_structures_lab/structures/graph.go
package structures

type Graph struct { edges map[string][]string }
func NewGraph()*Graph{return &Graph{edges:map[string][]string{}}}
func (g *Graph) AddVertex(v string){if _,ok:=g.edges[v];!ok{g.edges[v]=nil}}
func (g *Graph) AddEdge(from,to string){g.AddVertex(from);g.AddVertex(to);g.edges[from]=append(g.edges[from],to)}
func (g *Graph) BFS(start string)[]string{if _,ok:=g.edges[start];!ok{return nil};seen:=map[string]bool{start:true};q:=[]string{start};out:=[]string{};for len(q)>0{v:=q[0];q=q[1:];out=append(out,v);for _,n:=range g.edges[v]{if !seen[n]{seen[n]=true;q=append(q,n)}}};return out}
func (g *Graph) DFS(start string)[]string{seen:=map[string]bool{};out:=[]string{};var walk func(string);walk=func(v string){if seen[v]{return};seen[v]=true;out=append(out,v);for _,n:=range g.edges[v]{walk(n)}};if _,ok:=g.edges[start];ok{walk(start)};return out}
